src/hooks/useGetLocationState.ts

```diff
@@ -0,0 +1,17 @@
+import { useLocation } from 'react-router-dom';
+/**
+ * @description useLocation을 사용해 전달된 state값을 가져오는 함수입니다
+ * @returns ex useGetLocationState<{partId:string}>();
+ * @example const useGetLocationState<{partId:string}>();
+ */
+function useGetLocationState(): { [ket: string]: string | undefined } {
+  const location = useLocation();
+  const state = location.state;
+  //state가 빈 객체일 경우
+  if (state === null) {
+    return {};
+  }
+  return { ...state };
+}
+
+export default useGetLocationState;
```

Comment:
```suggestion
function useGetLocationState<T extends Record<PropertyKey, any>>() {
  return (useLocation()?.state ?? {}) as T
}
export default useGetLocationState;
```
@rhehfl @bluetree7878 
1. 현재 `js-doc example` 케이스를 보면 `제네릭`을 허용하는 것 같은데, 현재 훅은 제네릭을 허용하고 있지 않습니다. `as`를 활용한 타입 단언으로 동작하고 있어 제네릭을 허용하도록 수정합니다. 😂
2. state를 전개연산자로 굳이 복사 할 필요가 없습니다. 
- location.state를 사용 케이스를 고려해보자면, 해당 값을 직접 바꾸는 케이스는 거의 없을 거라 생각됩니다 :) 

아래는 사용 사례입니다.
```ts
const state = useGetLocationState<{ partId: string }>()
// const state: { partId: string }
```

Reply:
@rhehfl @bluetree7878 
해당 훅의 의도는 이해가 되지만 사실 location state를 위한 유틸 타입 하나를 추가하면 이런 `useGetLocationState`라는 훅을 추가해서 사용 할 필요가 없어집니다.

아래와 같이 사용해보는건 어떻게 생각하실까요?
```ts
export interface LocationState<T>{
  state: T
}

const { state } = useLocation() as LocationState<{ partId: string}>
```

Reply:
리뷰 감사합니다!
일단 1번 케이스의 경우 제가 훅을 수정하며 `js-doc example`을 깜빡하고 수정하지 않은 것 같습니다😓
2번 부분에서  `Record`유틸리티 타입에 `PropertyKey`으로 객체 키의 타입을 지정할수 있다는거는 몰랐네요
`type PropertyKey = string | number | symbol;`
객체에 키 타입을 지정할때 굉장히 유용할 것 같습니다 
@ssi02014 님 말씀대로 `??`연산자로 더 깔끔하게 코드를 작성할 수 있을 것 같습니다 감사합니다!!

Reply:
추상화에 관한 의견 감사합니다! 읽고 추상화에 대해 계속 생각을 해보느라 답이 늦어졌네요..

이 훅에 대해서 조금 생각을 해봤는데요 어떤게 더 좋을까에 대해 생각을 계속 해보니 `useGetLocationState`라는 훅으로 추상화를 시켜도 어떤 역할을 하는지는 대강 알 수 있지만 재사용성의 측면에서는 아직은 굳이 훅으로 분류할 필요까지는 없을 것 같습니다.

또한 훅 내부의 코드 길이도 긴 편이 아니기에 @ssi02014 님이 제공해주신 코드처럼 유틸 타입을 추가하여 사용하는 방법으로도 `Quiz`컴포넌트의 코드를 읽을때 훅으로 감싼거보다는 더욱 가독성이 좋을 것 같다는 생각이 들었습니다!

결론적으로는 굳이 `useGetLocationState`라는 훅을 만들지 않아도 `useLocation`라는 훅 자체가 현재 라우팅된 컴포넌트에 대한 정보 (경로, state 등) 을 담고있는 추상화된 훅이라고 생각해서
 `const { state } = useLocation() as LocationState<{ partId: string}>` 으로 타입을 지정해주는 것 만으로 충분하다고 생각이 드네요!
리뷰 감사합니다!!!

Reply:
``` ts
const { partId } = useLocation().state as { partId: number };
```
혹시 인터페이스를 만들지 않고 이런식으로 바로 타입 단언을 해주는 방식에 대해서는 어떻게 생각하시나요??
제 생각으로는 해당 컴포넌트 자체는 `learn`페이지에서 `partId` 를 전달해주기 때문에 바로 state에 접근해도 괜찮을 것 같다는 생각이 들고` partId `자체의 값이`"a" ,undefined` 와 같은 값이 올 일은 없다고 생각하기 때문에 타입 단언을 해주어도 괜찮을 것 같다는 생각이 듭니다!

Reply:
제 생각에도 @rhehfl 님께서 예시를 들어준 것처럼 타입 단언을 해도 가독성면으로도 괜찮고 딱히 큰 문제가 없을 것 같고, 가독성면으로도 괜찮은 것 같습니다. 하나 궁금한 게 위에서 partId의 타입이 `string`으로 계속 선언을 했었는데, 지금은 `number`로 바꾸신 게 원래 `number`가 맞았던 거죠?

Reply:
넵 어디서 착각을 했는지는 모르겠으나🙄 `useLocation`에 전달되는 state가 `string`형태로 전달될거라고 착각을 하고 있었던 것 같습니다.
확인결과 `partId : number`형태로 잘 전달되는것 같아서 수정했습니다!

Reply:
@rhehfl 넵 말씀주신 부분도 좋습니다 :) 추상화 여부가 메인 논의 주제였기 때문에 현재 상태도 좋습니다